libraries/AP_Compass/AP_Compass_IST8308.cpp

```diff
@@ -0,0 +1,246 @@
+        goto fail;
+    }
+
+    // DRDY enabled
+    // Dynamic Range=±500 uT, Sensitivity=6.6 LSB/uT
+    // OSR 16 (max 100Hz)
+    // Start continuous mode at 100Hz
+    if (!_dev->write_register(CNTL3_REG, CNTL3_VAL_DRDY_EN) ||
+        !_dev->write_register(CNTL4_REG, CNTL4_VAL_DYNAMIC_RANGE_500) ||
+        !_dev->write_register(OSRCNTL_REG, OSRCNTL_VAL_Y_16 | OSRCNTL_VAL_XZ_16) ||
+        !_dev->write_register(CNTL2_REG, CNTL2_VAL_CONT_ODR100_MODE)) {
+        printf("IST8308: found device but could not set it up\n");
+        goto fail;
+    }
+
+    // lower retries for run
+    _dev->set_retries(3);
+
+    _dev->get_semaphore()->give();
+
+    _instance = register_compass();
+
+    printf("%s found on bus %u id %u address 0x%02x\n", name,
+           _dev->bus_num(), _dev->get_bus_id(), _dev->get_bus_address());
+
+    set_rotation(_instance, _rotation);
+
+    _dev->set_device_type(DEVTYPE_IST8308);
+    set_dev_id(_instance, _dev->get_bus_id());
+
+    if (_force_external) {
+        set_external(_instance, true);
+    }
+
+    _periodic_handle = _dev->register_periodic_callback(SAMPLING_PERIOD_USEC,
+        FUNCTOR_BIND_MEMBER(&AP_Compass_IST8308::timer, void));
+
+    _perf_xfer_err = hal.util->perf_alloc(AP_HAL::Util::PC_COUNT, "IST8308_xfer_err");
+
+    return true;
+
+fail:
+    _dev->get_semaphore()->give();
+    return false;
+}
+
+void AP_Compass_IST8308::timer()
+{
+    struct PACKED {
+        le16_t rx;
+        le16_t ry;
+        le16_t rz;
+    } buffer;
+    uint8_t stat;
+
+    if (!_dev->read_registers(STAT1_REG, &stat, 1) ||
+        !(stat & STAT1_VAL_DRDY)) {
+        hal.util->perf_count(_perf_xfer_err);
+        return;
+    }
+
+    if (stat & STAT1_VAL_DOR) {
+        printf("IST8308: data overrun\n");
+    }
+
+    if (!_dev->read_registers(DATAX_L_REG, (uint8_t *) &buffer,
+                              sizeof(buffer))) {
+        hal.util->perf_count(_perf_xfer_err);
+        return;
+    }
+
+    auto x = static_cast<int16_t>(le16toh(buffer.rx));
+    auto y = static_cast<int16_t>(le16toh(buffer.ry));
+    auto z = static_cast<int16_t>(le16toh(buffer.rz));
+
+    // flip Z to conform to right-hand rule convention
+    z = -z;
+
+    /* Resolution: 0.1515 µT/LSB - already convert to milligauss */
+    Vector3f field = Vector3f{x * 1.515f, y * 1.515f, z * 1.515f};
+
+    /* rotate raw_field from sensor frame to body frame */
+    rotate_field(field, _instance);
+
+    /* publish raw_field (uncorrected point sample) for calibration use */
+    publish_raw_field(field, _instance);
+
+    /* correct raw_field for known errors */
+    correct_field(field, _instance);
+
+    if (_sem->take(HAL_SEMAPHORE_BLOCK_FOREVER)) {
+        _accum += field;
+        _accum_count++;
+        _sem->give();
+    }
+}
+
+void AP_Compass_IST8308::read()
+{
+    if (!_sem->take_nonblocking()) {
+        return;
+    }
+
+    if (_accum_count == 0) {
+        _sem->give();
+        return;
+    }
+
+    Vector3f field(_accum);
```

Comment:
Is there any reason to use another Vector3f instead of operating on `_accum`?

Reply:
this comes from other drivers that do the same thing. Looking again, the reasons seems to be lost to history: to be able to call `publish_filtered_field()` without holding the semaphore. Checking what this function does it doesn't appear to be needed.

Reply:
done